README
Module::Data



INSTALLATION

This is a Perl module distribution. It should be installed with whichever
tool you use to manage your installation of Perl, e.g. any of

  cpanm .
  cpan  .
  cpanp -i .

Consult http://www.cpan.org/modules/INSTALL.html for further instruction.
Should you wish to install this module manually, the procedure is

  perl Build.PL
  ./Build
  ./Build test
  ./Build install

COPYRIGHT AND LICENSE

This software is copyright (c) 2017 by Kent Fredric <kentnl@cpan.org>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Release history for Module-Data

0.014 2017-03-06T11:55:33Z 95f66c8

0.013 2017-03-06T11:43:10Z 095bcf3
 [Bugfix]
 - Avoid test failures with -Ddefault_inc_excludes_dot
 - Closes https://github.com/kentnl/Module-Data/issues/2

 [Dependencies::Stats]
 - Dependencies changed since 0.012, see misc/*.deps* for details
 - build: -1
 - configure: +2 -1 (recommends: +1)
 - develop: +24 ↑3 -14 (recommends: +1, suggests: +2 -1)
 - runtime: (recommends: +2)
 - test: +3 ↓1 (recommends: +3)

 [Tooling]
 - Switched from Module::Build to EUMM. WARNING: This could break some vendors.

0.012 2013-11-25T08:27:23Z
 [Bugfix]
 - Don't try to construct a path if underlying mechanisms returned no path.

0.011 2013-11-25T02:13:53Z
 [Dependencies::Noteworthy]
 - Require newer Path::ScanINC

 [Dependencies::Stats]
 - Dependencies changed since 0.010, see misc/*.deps* for details
 - develop: ↑1
 - runtime: ↑1

 [Maintenance]
 - Changelog retroactively tidied.

0.010 2013-11-21T12:25:03Z
 [00 Minor]
 - Internals reimplemented in Path::Tiny
 - Trial release to give end users a bit of warning, because Path::Class objects were previously emitted.

 [Dependencies::Noteworthy]
 - Requires Path::Tiny
 - No longer requires Path::Class

 [Dependencies::Stats]
 - Dependencies changed since 0.007, see misc/*.deps* for details
 - runtime: +1 -1
 - test: -1

0.007 2013-11-20T13:09:23Z
 [00 BugFix]
 - Prevent test failure with new Path::ScanINC

 [Dependencies::Noteworthy]
 - require newer Module::Build
 - require newer Test::More

 [Dependencies::Stats]
 - Dependencies changed since 0.006, see misc/*.deps* for details
 - build: ↑1
 - configure: ↑1
 - develop: +50 -1 (recommends: -1, suggests: +1 -1)
 - test: ↑1 -2

 [Documentation]
 - Add README.mkdn
 - Update ©
 - use UTF∞8

 [Meta]
 - Bugtracker to GH Issues

 [Packaging]
 - Use test_requires

 [Tests]
 - use Test::Compile::PerFile
 - Update ReportVersions::Tiny
 - Update CPAN::Changes
 - Update Test::Kwalitee

0.006 2012-04-13T04:02:46Z
 [Tests]
 - Replace call to Data::Dump::pp with a hand-written simple dumper as pp is for some bizzare reason entering
   recursion.

0.005 2012-04-11T20:33:42Z
 [Bugfixes]
 - Update of the test code to not fail prematurely under Test::Builder2

 [Tests]
 - Refactored Whitelisting logic into its own t/tlib/ module 'Whitelist'
 - Updated 02 & 03 to whitelist modules loaded by Test::Builder2

0.004 2012-04-11T06:01:11Z
 [Dependencies::Noteworthy]
 - Path::ScanINC is now 0.002 minimum

 [Dependencies::Stats]
 - Dependencies changed since 0.003, see misc/*.deps* for details
 - runtime: ↑1

0.003 2012-04-10T09:11:18Z
 [Dependencies::New]
 - Test requires: Carp

 [Dependencies::Stats]
 - Dependencies changed since 0.002, see misc/*.deps* for details
 - test: +1

 [Tests]
 - Improved 03_fake_system.t reporting to reveal why win32 is requring() "wrong".

0.002 2012-04-09T02:01:46Z
 [Dependencies::Noteworthy]
 - uses Module::Metadata
 - uses Path::ScanINC
 - uses Path::Class

 [Dependencies::Stats]
 - Dependencies changed since 0.001, see misc/*.deps* for details
 - runtime: +2 -1
 - test: +2

 [Internals]
 - Rework module discovery to be optimistic, check %INC if its there already, and if it is, report where it was found,
   otherwise, ask Path::ScanINC to tell us where it is WITHOUT loading it. This will hopefully be usefull in many cases
   allowing us to detect module availability without needing the compile/memory overheads of actually loading said
   modules
 - Rework ->version call to be optimistic, ask PACKAGE->VERSION for version if the package appears to be loaded,
   otherwise, fall back on Path::ScanINC to find the file, and Module::Metadata to extract the version out of it. The
   hope is that will eliminate the need to load the whole module permenantly into memory *just* to discern what version
   it is.

0.001 2012-03-23T15:37:57Z
 - First version.


t/tlib/Whitelist.pm
use strict;
use warnings;

package Whitelist;

# FILENAME: Whitelist.pm
# CREATED: 12/04/12 06:33:42 by Kent Fredric (kentnl) <kentfredric@gmail.com>
# ABSTRACT: A basic test whitelist
use Carp qw( confess );

sub new {
	my ( $class, @args ) = @_;
	return bless { args => \@args }, $class;
}

sub whitelist {
	my ( $self, @args ) = @_;
	die 'Frozen' if $self->{frozen};
	$self->{whitelist} = [] unless exists $self->{whitelist};
	$self->{load}      = [] unless exists $self->{load};

	push @{ $self->{whitelist} }, @args;
	push @{ $self->{load} },      @args;

}

sub noload_whitelist {
	my ( $self, @args ) = @_;
	die 'Frozen' if $self->{frozen};
	$self->{whitelist} = [] unless exists $self->{whitelist};
	push @{ $self->{whitelist} }, @args;
}

sub freeze {
	my ($self) = @_;
	$self->{module_whitelist} = {};
	$self->{whitelist_inc}    = {};
	require Module::Runtime;
	for my $lib ( @{ $self->{load} } ) {
		Module::Runtime::require_module($lib);
	}
	for my $lib ( @{ $self->{whitelist} } ) {
		my $nn = Module::Runtime::module_notional_filename($lib);
		$self->{module_whitelist}->{$nn} = 1;
		$self->{whitelist_inc}->{$nn} = $INC{$nn} if exists $INC{$nn};
	}
	$self->{frozen}   = 1;
	$self->{real_inc} = {%INC};
}

sub checker {
	my ($self) = @_;
	my $dumper = sub {
		my $dump_hash = sub {
			my $hash   = shift;
			my $prefix = shift;
			$prefix = "  " if not defined $prefix;
			return qq[{\n$prefix] . ( join qq{,\n$prefix}, map { $_ . ' => ' . $hash->{$_} } keys %{$hash} ) . qq[\n}];
		};
		my $string_values = sub {
			my $hash = shift;
			return { map { $_, 'q{' . $hash->{$_} . '}' } keys %{$hash} };
		};

		return $dump_hash->(
			{
				real_inc         => $dump_hash->( $string_values->( $self->{real_inc}, ),         "    " ),
				whitelist_inc    => $dump_hash->( $string_values->( $self->{whitelist_inc}, ),    "    " ),
				module_whitelist => $dump_hash->( $string_values->( $self->{module_whitelist}, ), "    " ),

			}
		);
	};

	my $coderef;
	$coderef = sub {
		my ( $code, $filename ) = @_;
		return if exists $self->{module_whitelist}->{$filename};
		confess( "$filename requested but not whitelisted:\n " . $dumper->() );
	};

}
1;



t/03_t/lib/VERSION/Test/A.pm
use strict;
use warnings;

package Test::A;

# FILENAME: A.pm
# CREATED: 26/03/12 14:03:00 by Kent Fredric (kentnl) <kentfredric@gmail.com>
# ABSTRACT: test package

our $VERSION = 1.00;

sub example {
	return "A(D)";
}

1;



t/03_t/lib/VERSION/Test/B.pm
use strict;
use warnings;

package Test::B;

# FILENAME: A.pm
# CREATED: 26/03/12 14:03:00 by Kent Fredric (kentnl) <kentfredric@gmail.com>
# ABSTRACT: test package

our $VERSION = 1.00;

sub example {
	return "B(D)";
}

1;



Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.009.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Introspect context information about modules in \@INC",
  "AUTHOR" => "Kent Fredric <kentnl\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Module-Data",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "Module::Data",
  "PREREQ_PM" => {
    "Module::Metadata" => 0,
    "Module::Runtime" => 0,
    "Moo" => 0,
    "Path::ScanINC" => "0.011",
    "Path::Tiny" => 0,
    "Sub::Quote" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Carp" => 0,
    "Data::Dump" => 0,
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "FindBin" => 0,
    "Test::Fatal" => 0,
    "Test::More" => "0.96"
  },
  "VERSION" => "0.014",
  "test" => {
    "TESTS" => "t/*.t t/00-compile/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "Data::Dump" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "FindBin" => 0,
  "Module::Metadata" => 0,
  "Module::Runtime" => 0,
  "Moo" => 0,
  "Path::ScanINC" => "0.011",
  "Path::Tiny" => 0,
  "Sub::Quote" => 0,
  "Test::Fatal" => 0,
  "Test::More" => "0.96",
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


lib/Module/Data.pm
use 5.006;    # our
use strict;
use warnings;

package Module::Data;

our $VERSION = '0.014';

# ABSTRACT: Introspect context information about modules in @INC

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY

use Moo qw( around has );
use Sub::Quote qw( quote_sub );

around BUILDARGS => sub {
  my ( $orig, $class, @args ) = @_;

  unshift @args, 'package' if @args % 2 == 1;

  return $class->$orig(@args);
};






































has package => (
  required => 1,
  is       => 'ro',
  ## no critic (ValuesAndExpressions::RequireInterpolationOfMetachars, ValuesAndExpressions::RestrictLongStrings)
  isa => quote_sub q{}
    . q{die "given undef for 'package' , expects a Str/module name" if not defined $_[0];}
    . q{die " ( 'package' => $_[0] ) is not a Str/module name, got a ref : " . ref $_[0] if ref $_[0];}
    . q{require Module::Runtime;}
    . q{Module::Runtime::check_module_name( $_[0] );},
);

has _notional_name => (
  is   => 'ro',
  lazy => 1,
  ## no critic (ValuesAndExpressions::RequireInterpolationOfMetachars)
  default => quote_sub q{} . q{require Module::Runtime;} . q{return Module::Runtime::module_notional_filename( $_[0]->package );},
);











sub loaded {
  my ($self) = @_;
  return exists $INC{ $self->_notional_name };
}















## no critic ( ProhibitBuiltinHomonyms )
sub require {
  my ($self) = @_;
  return $self->package if $self->loaded;

  require Module::Runtime;
  Module::Runtime::require_module( $self->package );
  return $self->package;
}

sub _find_module_perl {
  my ($self) = @_;
  $self->require;
  return $INC{ $self->_notional_name };
}

sub _find_module_emulate {
  my ($self) = @_;
  require Path::ScanINC;
  Path::ScanINC->VERSION('0.011');
  return Path::ScanINC->new()->first_file( $self->_notional_name );
}

sub _find_module_optimistic {
  my ($self) = @_;
  return $INC{ $self->_notional_name } if $self->loaded;
  return $self->_find_module_emulate;
}

## use critic















has path => (
  is       => 'ro',
  lazy     => 1,
  init_arg => undef,
  builder  => '_build_path',
);

sub _build_path {
  my ( $self, ) = @_;
  my $value = $self->_find_module_optimistic;
  return if not defined $value;
  require Path::Tiny;
  return Path::Tiny::path($value)->absolute;
}

















has root => (
  is       => 'ro',
  lazy     => 1,
  init_arg => undef,
  builder  => '_build_root',
);

sub _build_root {
  my ($path) = $_[0]->path;

  # Parent ne Self is the only cross-platform way
  # I can think of that will stop at the top of a tree
  # as / is not applicable on windows.
  while ( $path->parent->absolute ne $path->absolute ) {
    if ( not $path->is_dir ) {
      $path = $path->parent;
      next;
    }
    if ( $path->child( $_[0]->_notional_name )->absolute eq $_[0]->path->absolute ) {
      return $path->absolute;
    }
    $path = $path->parent;
  }
  return;

}





















sub _version_perl {
  my ($self) = @_;
  $self->require;

  # has to load the code into memory to work
  return $self->package->VERSION;
}

sub _version_emulate {
  my ($self) = @_;
  my $path = $self->path;
  require Module::Metadata;
  my $i = Module::Metadata->new_from_file( $path, collect_pod => 0 );
  return $i->version( $self->package );
}

sub _version_optimistic {
  my ($self) = @_;
  return $self->package->VERSION if $self->loaded;
  return $self->_version_emulate;
}

sub version {
  my ( $self, ) = @_;
  return $self->_version_optimistic;
}









no Moo;

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Module::Data - Introspect context information about modules in @INC

=head1 VERSION

version 0.014

=head1 SYNOPSIS

	use Module::Data;

	my $d = Module::Data->new( 'Package::Stash' );

	$d->path; # returns the path to where Package::Stash was found in @INC

	$d->root; # returns the root directory in @INC that 'Package::Stash' was found inside.

	# Convenient trick to discern if you're in a development environment

	my $d = Module::Data->new( 'Module::Im::Developing' );

	if ( -e $d->root->parent->subdir('share') ) {
		# Yep, this dir exists, so we're in a dev context.
		# because we know in the development context all modules are in lib/*/*
		# so if the modules are anywhere else, its not a dev context.
		# see File::ShareDir::ProjectDistDir for more.
	}

	# Helpful sugar.

	my $v = $d->version;

=head1 METHODS

=head2 package

Returns the package the C<Module::Data> instance was created for. In essence,
this will just return the value you passed during C<new>, nothing more, nothing
less.

	my $package = $md->package

=head2 loaded

Check to see if the module is already recorded as being loaded in C<%INC>

	if ( $md->loaded ) {
		say "$md was loaded";
	}

=head2 require

Require the module be loaded into memory and the global stash.

  my $mod = Module::Data->new( 'Foo' ); # nothing much happens.
  $mod->require; # like 'require Foo';

Returns the L</package> name itself for convenience so you can do

  my $mod = Module::Data->new('Foo');
  $mod->require->new( %args );

=head2 path

A Path::Tiny object with the absolute path to the found module.

	my $md = Module::Data->new( 'Foo' );
	my $path = $md->path;

C<$path> is computed optimistically. If the L</package> is listed as being
L</loaded>, then it asks C<%INC> for where it was found, otherwise, the path is
resolved by simulating C<perl>'s path look up in C<@INC> via
L<< C<Path::ScanINC>|Path::ScanINC >>.

=head2 root

Returns the base directory of the tree the module was found at.
( Probably from @INC );

	local @INC = (
		"somewhere/asinine/",
		"somewhere/in/space/",   # Where Lib::Foo::Bar is
		"somethingelse/",
	);
	my $md = Module::Data->new( "Lib::Foo::Bar");
	$md->path ; # somewhere/in/space/Lib/Foo/Bar.pm
	my $root = $md->root # somewhere/in/space

=head2 version

If the module appears to be already loaded in memory:

	my $v = $md->version;

is merely shorthand for $package->VERSION;

However, if the module is not loaded into memory, all efforts to extract the
value without loading the code permanently are performed.

Here, this means we compute the path to the file manually ( see L</path> ) and
parse the file with L<< C<Module::Metadata>|Module::Metadata >> to statically extract C<$VERSION>.

This means you can unleash this code on your entire installed module tree, while
incurring no permanent memory gain as you would normally incur if you were to
C<require> them all.

=for Pod::Coverage   BUILDARGS

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2017 by Kent Fredric <kentnl@cpan.org>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


maint/perlcritic.rc.gen.pl
#!/usr/bin/env perl
# FILENAME: bundle_to_ini.pl
# CREATED: 02/06/14 01:48:56 by Kent Fredric (kentnl) <kentfredric@gmail.com>
# ABSTRACT: Write an INI file from a bundle

use 5.008;    # utf8
use strict;
use warnings;
use utf8;

our $VERSION = 0.001;

use Carp qw( croak carp );
use Perl::Critic::ProfileCompiler::Util qw( create_bundle );
use Path::Tiny qw(path);

## no critic (ErrorHandling::RequireUseOfExceptions)
my $bundle = create_bundle('Example::Author::KENTNL');
$bundle->configure;

my @stopwords = (
);
for my $var (@stopwords) {
  $bundle->add_or_append_policy_field( 'Documentation::PodSpelling' => ( 'stop_words' => $var ) );
}
#$bundle->add_or_append_policy_field(
#  'Subroutines::ProhibitCallsToUndeclaredSubs' => ( 'exempt_subs' => 'String::Formatter::str_rf' ), );

$bundle->remove_policy('Subroutines::ProhibitCallsToUndeclaredSubs');
$bundle->remove_policy('Variables::ProhibitUnusedVarsStricter');
#$bundle->remove_policy('ErrorHandling::RequireCarping');
#$bundle->remove_policy('NamingConventions::Capitalization');

my $inf = $bundle->actionlist->get_inflated;

my $config = $inf->apply_config;

{
  my $rcfile = path('./perlcritic.rc')->openw_utf8;
  $rcfile->print( $config->as_ini, "\n" );
  close $rcfile or croak 'Something fubared closing perlcritic.rc';
}
my $deps = $inf->own_deps;
{
  my $target = path('./misc');
  $target->mkpath if not $target->is_dir;

  my $depsfile = $target->child('perlcritic.deps')->openw_utf8;
  for my $key ( sort keys %{$deps} ) {
    $depsfile->printf( "%s~%s\n", $key, $deps->{$key} );
    *STDERR->printf( "%s => %s\n", $key, $deps->{$key} );
  }
  close $depsfile or carp 'Something fubared closing perlcritic.deps';
}



Build.PL

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v6.009.
use strict;
use warnings;

use Module::Build 0.28;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.4220"
  },
  "configure_requires" => {
    "Module::Build" => "0.4220",
    "perl" => "5.006"
  },
  "dist_abstract" => "Introspect context information about modules in \@INC",
  "dist_author" => [
    "Kent Fredric <kentnl\@cpan.org>"
  ],
  "dist_name" => "Module-Data",
  "dist_version" => "0.013",
  "license" => "perl",
  "module_name" => "Module::Data",
  "recommends" => {
    "Moo" => "1.000008",
    "Path::Tiny" => "0.058"
  },
  "recursive_test_files" => 1,
  "requires" => {
    "Module::Metadata" => 0,
    "Module::Runtime" => 0,
    "Moo" => 0,
    "Path::ScanINC" => "0.011",
    "Path::Tiny" => 0,
    "Sub::Quote" => 0,
    "perl" => "5.006",
    "strict" => 0,
    "warnings" => 0
  },
  "test_requires" => {
    "Carp" => 0,
    "Data::Dump" => 0,
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "FindBin" => 0,
    "Test::Fatal" => 0,
    "Test::More" => "1.302073",
    "perl" => "5.006"
  }
);


my %fallback_build_requires = (
  "Carp" => 0,
  "Data::Dump" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "FindBin" => 0,
  "Module::Build" => "0.4220",
  "Test::Fatal" => 0,
  "Test::More" => "1.302073",
  "perl" => "5.006"
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);


$build->create_build_script;
